shared/src/proto/types.rs

```diff
@@ -51,54 +52,47 @@ pub struct SignedTxData {
     pub sig: common::Signature,
 }
 
-/// A generic signed data wrapper for Borsh encode-able data.
+/// Tag type that indicates we should use [`BorshSerialize`]
+/// to sign data in a [`Signed`] wrapper.
+#[derive(Eq, PartialEq, Clone, Debug, Serialize, Deserialize)]
+pub enum SerializeWithBorsh {}
+
+/// A generic signed data wrapper for serialize-able types.
+///
+/// The default serialization method is [`BorshSerialize`].
 #[derive(
-    Clone, Debug, BorshSerialize, BorshDeserialize, Serialize, Deserialize,
+    Eq, Clone, Debug, BorshSerialize, BorshDeserialize, Serialize, Deserialize,
```

Comment:
I'm guessing `Eq` should only be derived for the case where `T: PartialEq`? It might need to be manually implemented in that case

Reply:
*`T: PartialEq + Eq`

Reply:
`Eq` doesn't add any methods, it just means we a have a full equivalence relation.
https://doc.rust-lang.org/stable/std/cmp/trait.Eq.html

Reply:
if you have `PartialEq` implemented, `Eq` can be derived

Reply:
I mean we can (or should?) only derive `Eq` in the case where `T: Eq`. As you can sign `T`s which don't implement `PartialEq`/`Eq`. I'm surprised the compiler lets you annotate `Eq` on `Signed<T, S>` without `T` also having the `Eq` trait bound though so I could be misunderstanding something

Reply:
something like

Reply:
the compiler already does that automatically. it'll only implement `Eq` if the type params are also `Eq`

Reply:
actually not true, it'll implement it if the type params are `PartialEq`. ehhhh
https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=095498c880c5653848415b9228642fb7

alright cool I'll manually implement `Eq`

Reply:
> the compiler already does that automatically. it'll only implement `Eq` if the type params are also `Eq`

Yeah I thought this is how it should work as well and still don't fully understand this. Like for `PartialEq` the compiler could only derive it if `T: PartialEq` it seems. Maybe will revisit this.